bindgen/clang.rs

```diff
@@ -555,36 +543,41 @@ impl Cursor {
 
         let (file, other_file) = match (file.name(), other_file.name()) {
             (Some(file), Some(other_file)) => (file, other_file),
-            // Built-in definitions should come first.
-            (Some(_), None) => return cmp::Ordering::Greater,
-            (None, Some(_)) => return cmp::Ordering::Less,
-            (None, None) => return cmp::Ordering::Equal,
+            // Keep the original sorting of built-in definitions.
+            (None, _) | (_, None) => return cmp::Ordering::Equal,
         };
 
+        // If both items are in the same source file, simply compare the offset.
         if file == other_file {
-            // Both items are in the same source file, compare by byte offset.
             return offset.cmp(&other_offset);
         }
 
+        // `None` here means `file`/`other_file` is the main header file.
         let include_location = ctx.included_file_location(&file);
         let other_include_location = ctx.included_file_location(&other_file);
+
         match (include_location, other_include_location) {
-            (Some((file2, offset2)), _) if file2 == other_file => {
+            // The main header file (`None`) comes after header passed as CLI argument (`Some((None, _))`).
+            (None, Some((None, _))) => cmp::Ordering::Greater,
+            (Some((None, _)), None) => cmp::Ordering::Less,
+            // If an item was included in the same source file as the other item,
+            // compare its `#include` location offset the offset of the other item.
+            (Some((Some(file2), offset2)), _) if file2 == other_file => {
                 offset2.cmp(&other_offset)
             }
-            (Some(_), None) => cmp::Ordering::Greater,
-            (_, Some((other_file2, other_offset2))) if file == other_file2 => {
+            (_, Some((Some(other_file2), other_offset2)))
+                if file == other_file2 =>
+            {
                 offset.cmp(&other_offset2)
             }
-            (None, Some(_)) => cmp::Ordering::Less,
-            (Some((file2, offset2)), Some((other_file2, other_offset2))) => {
-                if file2 == other_file2 {
-                    offset2.cmp(&other_offset2)
-                } else {
-                    cmp::Ordering::Equal
-                }
+            // If both items were included in the same file, compare the offset of their `#include` directives.
+            (Some((file2, offset2)), Some((other_file2, other_offset2)))
+                if file2 == other_file2 =>
+            {
+                offset2.cmp(&other_offset2)
```

Comment:
I don't think this is enough. You need to somehow deal with transitive includes don't you?

Reply:
Can you give an example for what you think might get sorted wrong? Will try to add a test for this then.

Note that when recursing, the recursive `visit_sorted` gets sorted in and of itself, so this may actually work without dealing with transitive includes explicitly.

Reply:
I don't have a concrete test-case because it would depend on the concrete sorting algorithm, but not dealing with recursive includes would mean that you break the `Ord` invariants, as in, you might have a, b, and c, such that `a < b`, `b < c`, but `a == c` or so, breaking https://doc.rust-lang.org/std/cmp/trait.Ord.html#corollaries, and potentially invariants of the sorting algorithm. Need to take a look at the new approach.

Reply:
I agree, we shouldn't rely on the fact that sort works for the testcases we have so far even though we don't have an actual ordering as it is not transitive. 

I'd be happy to fix this in an upcoming PR if that's ok with both of you.